File: languages/c/gemm-par.c
// gemm-par: parallel (wall-clock scaling track) version of gemm. Produces the SAME
// result as serial gemm.c for ANY core count, so it passes the same checksum.
// Invocation:  gemm-par <cores> <n>
// Row-band decomposition: worker w computes output rows [w*N/cores, (w+1)*N/cores)
// with the SAME pinned i,k,j loop order as serial; each worker writes only its own
// rows of C (disjoint, no shared writes) -> result is bit-identical to serial.
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

// COMPUTE_NS marker (to stderr): elapsed of the parallel compute region only, so the
// scaling track's speedup excludes data-gen + runtime startup. measure-scaling.sh reads it.
static long long now_ns(void) { struct timespec t; clock_gettime(CLOCK_MONOTONIC, &t); return (long long)t.tv_sec * 1000000000LL + t.tv_nsec; }

#define P 1000000007L

static long lcg(long s) { return (s * 1103515245L + 12345L) & 0x7fffffffL; }

static int N;
static long *A, *B, *C;

typedef struct { int lo, hi; } Band;

static void *worker(void *arg) {
    Band *b = (Band *)arg;
    for (int i = b->lo; i < b->hi; i++) {
        for (int k = 0; k < N; k++) {
            long a = A[i * N + k];
            for (int j = 0; j < N; j++) {
                C[i * N + j] += a * B[k * N + j];
            }
        }
    }
    return NULL;
}

int main(int argc, char **argv) {
    int cores = argc > 1 ? atoi(argv[1]) : 1;
    N = argc > 2 ? atoi(argv[2]) : 256;
    if (cores < 1) cores = 1;
    if (cores > N) cores = N;
    if (cores > 256) cores = 256;

    A = malloc((size_t)N * N * sizeof(long));
    B = malloc((size_t)N * N * sizeof(long));
    C = calloc((size_t)N * N, sizeof(long));

    // Sequential init, identical LCG sequence to serial -> identical A, B, C.
    long s = 42;
    for (int i = 0; i < N * N; i++) { s = lcg(s); A[i] = s % 128; }
    for (int i = 0; i < N * N; i++) { s = lcg(s); B[i] = s % 128; }

    long long _t0 = now_ns();
    pthread_t th[256];
    Band bands[256];
    for (int w = 0; w < cores; w++) {
        bands[w].lo = (int)((long)w * N / cores);
        bands[w].hi = (int)((long)(w + 1) * N / cores);
        pthread_create(&th[w], NULL, worker, &bands[w]);
    }
    for (int w = 0; w < cores; w++) pthread_join(th[w], NULL);
    fprintf(stderr, "COMPUTE_NS %lld\n", now_ns() - _t0);

    long h = 0;
    for (int i = 0; i < N * N; i++) h = (h * 31 + C[i] % P) % P;
    long secondary = C[N * N - 1] % P;
    printf("%ld\n", h);
    printf("gemm(%d) = %ld\n", N, secondary);

    free(A); free(B); free(C);
    return 0;
}

File: languages/elixir/gemm-par.exs
# gemm-par: parallel scaling-track variant.
# Invocation: elixir gemm-par.exs <cores> <n>
# Decomposition: partition the N output rows of C into `cores` contiguous bands.
# Worker w computes rows [w*N/cores, (w+1)*N/cores). Loop order i->k->j is pinned
# (identical to serial). Each worker writes only its own rows of C (:atomics, disjoint).
# B is read-only and shared; concurrent reads are safe.
# Core-invariant: C[i*N+j] = sum_k A[i*N+k]*B[k*N+j], independent of core count.
# Checksum computed serially over full C row-major after Task.await_many.
# Output byte-identical to serial gemm for cores=1,2,4.
import Bitwise

defmodule GemmPar do
  @p 1_000_000_007

  defp lcg(s), do: band(s * 1_103_515_245 + 12_345, 0x7FFFFFFF)

  defp aget(ref, i), do: :atomics.get(ref, i + 1)
  defp aput(ref, i, v), do: :atomics.put(ref, i + 1, v)
  defp aadd(ref, i, v), do: :atomics.add(ref, i + 1, v)

  # Fill an :atomics array with values in 0..127 from the LCG stream.
  defp fill(_ref, i, total, state) when i >= total, do: state

  defp fill(ref, i, total, state) do
    state = lcg(state)
    aput(ref, i, rem(state, 128))
    fill(ref, i + 1, total, state)
  end

  # Inner j-loop: C[base+j] += a * B[kn+j]
  defp inner_j(_b, _c, _kn, _base, _a, j, n) when j >= n, do: :ok

  defp inner_j(b, c, kn, base, a, j, n) do
    aadd(c, base + j, a * aget(b, kn + j))
    inner_j(b, c, kn, base, a, j + 1, n)
  end

  # Middle k-loop
  defp mid_k(_a, _b, _c, _i, k, n) when k >= n, do: :ok

  defp mid_k(a, b, c, i, k, n) do
    av = aget(a, i * n + k)
    inner_j(b, c, k * n, i * n, av, 0, n)
    mid_k(a, b, c, i, k + 1, n)
  end

  # Outer i-loop over a row band [i_start, i_end).
  defp band_rows(_a, _b, _c, i, i_end, _n) when i >= i_end, do: :ok

  defp band_rows(a, b, c, i, i_end, n) do
    mid_k(a, b, c, i, 0, n)
    band_rows(a, b, c, i + 1, i_end, n)
  end

  # Checksum poly-hash over C in row-major order.
  defp hash_c(_c, i, total, h) when i >= total, do: h

  defp hash_c(c, i, total, h) do
    hash_c(c, i + 1, total, rem(h * 31 + rem(aget(c, i), @p), @p))
  end

  def run(cores, n) do
    nn = n * n
    a = :atomics.new(nn, signed: true)
    b = :atomics.new(nn, signed: true)
    c = :atomics.new(nn, signed: true)

    # Generate A then B with same LCG as serial gemm (seed=42).
    state = fill(a, 0, nn, 42)
    _state = fill(b, 0, nn, state)

    # C is already zero-initialised by :atomics.new.

    t0 = System.monotonic_time(:nanosecond)

    # Spawn one task per band; each writes only its own rows of C.
    tasks =
      Enum.map(0..(cores - 1), fn w ->
        i_start = div(w * n, cores)
        i_end = div((w + 1) * n, cores)
        Task.async(fn -> band_rows(a, b, c, i_start, i_end, n) end)
      end)

    # Barrier: wait for all workers to finish.
    Enum.each(tasks, fn task -> Task.await(task, :infinity) end)

    t1 = System.monotonic_time(:nanosecond)
    IO.puts(:stderr, "COMPUTE_NS #{t1 - t0}")

    # Serial checksum pass — identical to gemm.exs.
    h = hash_c(c, 0, nn, 0)
    secondary = rem(aget(c, nn - 1), @p)
    {h, secondary}
  end
end

[cores_s, n_s] =
  case System.argv() do
    [c, n | _] -> [c, n]
    [c] -> [c, "256"]
    _ -> ["1", "256"]
  end

cores = String.to_integer(cores_s)
n = String.to_integer(n_s)

{h, sec} = GemmPar.run(cores, n)
IO.puts(h)
IO.puts("gemm(#{n}) = #{sec}")
